File: spring/core/SpringBeanLifecycles/src/main/java/demos/lifecycle/prototype/scope/SampleShop.java
package demos.lifecycle.prototype.scope;

import demos.lifecycle.common.PaymentEngine;
import demos.lifecycle.common.PricingEngine;
import demos.lifecycle.common.Shop;
import demos.lifecycle.common.StockCheckEngine;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.context.annotation.Scope;
import org.springframework.stereotype.Service;

@Service("shop")
@Scope("prototype")
public class SampleShop implements Shop {
    private final PricingEngine pricingEngine;
    private final StockCheckEngine stockCheckEngine;
    private final PaymentEngine paymentEngine;

    @Autowired
    public SampleShop(@Qualifier("pricing") PricingEngine pricingEngine,
                      @Qualifier("payment") PaymentEngine paymentEngine,
                      @Qualifier("stock") StockCheckEngine stockCheckEngine) {
        super();
        this.pricingEngine = pricingEngine;
        this.paymentEngine = paymentEngine;
        this.stockCheckEngine = stockCheckEngine;
    }

    @Override
    public boolean makePurchase(String itemNo, int quantity, String cardNo) {
        if (stockCheckEngine.check(itemNo) >= quantity) {
            double charge = pricingEngine.price(itemNo, quantity);
            if (paymentEngine.authorize(cardNo, charge)) {
                return true;
            }
        }
        return false;
    }

    public PricingEngine getPricingEngine() {
        return pricingEngine;
    }

    public StockCheckEngine getStockCheckEngine() {
        return stockCheckEngine;
    }

    public PaymentEngine getPaymentEngine() {
        return paymentEngine;
    }
}

File: spring/core/AnnotationDrivenDIWithKotlin/src/main/kotlin/demos/spring/notes/v9/SampleShop.kt
package demos.spring.notes.v9

import demos.spring.notes.common.PaymentEngine
import demos.spring.notes.common.PricingEngine
import demos.spring.notes.common.Shop
import demos.spring.notes.common.StockCheckEngine
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.context.annotation.Scope
import org.springframework.stereotype.Service

@Service("shop")
@Scope("prototype")
class SampleShop @Autowired constructor(
        @Qualifier("pricing") val pricingEngine: PricingEngine,
        @Qualifier("stock") val stockCheckEngine: StockCheckEngine,
        @Qualifier("payment") val paymentEngine: PaymentEngine): Shop {

    override fun makePurchase(itemNo: String, quantity: Int, cardNo: String): Boolean {
        if (stockCheckEngine.check(itemNo) >= quantity) {
            val charge = pricingEngine.price(itemNo, quantity)
            if (paymentEngine.authorize(cardNo, charge)) {
                return true
            }
        }
        return false
    }
}
